README.md
A library for Delaunay triangulation for Dart developers.

This implementation is adapted from the [Delaunator][delaunator] JavaScript
library.

## Usage

A simple usage example:

```dart
Float32List points = Float32List.fromList(<double>[
  143.0, 178.5,
  50.2, -100.7,
  ...
]);
Delaunay delaunay = Delaunay(points);
delaunay.update();
for (int i = 0; i < delaunay.triangles.length; i += 3) {
  int a = delaunay.triangles[i];
  int b = delaunay.triangles[i + 1];
  int c = delaunay.triangles[i + 2];

  double ax = delaunay.coords[2*a];
  double ay = delaunay.coords[2*a + 1];
  ...
}
...
points[0] = 140.0;
delaunay.update();
...
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[delaunator]: https://github.com/mapbox/delaunator
[tracker]: https://github.com/zanderso/delaunay.dart/issues


CHANGELOG.md
## 2.0.0

- Updated SDK constraint and minor fixes.

## 1.0.0

- Migrated to null-safety.

## 0.1.1

- Version bump for pub.

## 0.1.0

- Initial version.


test/delaunay_test.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math';
import 'dart:typed_data';

import 'package:delaunay/delaunay.dart';
import 'package:test/test.dart';

import 'fixtures/ukraine.dart' as ukraine;

void main() {
  test('triangulates points and typed data', () {
    final Delaunay pointsDelaunay = Delaunay.from(ukraine.points)..update();
    final Delaunay dataDelaunay = Delaunay(ukraine.typedPoints)..update();
    expect(pointsDelaunay.triangles, equals(dataDelaunay.triangles));
  });

  test('produces correct triangulation', () {
    validate(ukraine.typedPoints);
  });

  test('produces correct triangulation after modifying coords in place', () {
    final Delaunay d = Delaunay(ukraine.typedPoints);

    validate(ukraine.typedPoints, d);

    expect(d.triangles.length, equals(5151));

    ukraine.typedPoints[0] = 80.0;
    ukraine.typedPoints[1] = 220.0;
    validate(ukraine.typedPoints, d);

    expect(d.triangles.length, equals(5157));
  });

  test('corner case 1', () {
    final Float32List data = Float32List.fromList(<double>[
      516,
      661,
      369,
      793,
      426,
      539,
      273,
      525,
      204,
      694,
      747,
      750,
      454,
      390,
    ]);
    validate(data);
  });

  test('corner case 2', () {
    final Float32List data = Float32List.fromList(<double>[
      382,
      302,
      382,
      328,
      382,
      205,
      623,
      175,
      382,
      188,
      382,
      284,
      623,
      87,
      623,
      341,
      141,
      227,
    ]);
    validate(data);
  });

  test('corner case 3', () {
    final Float32List data = Float32List.fromList(<double>[
      4,
      1,
      3.7974166882130675,
      2.0837249985614585,
      3.2170267516619773,
      3.0210869309396715,
      2.337215067329615,
      3.685489874065187,
      1.276805078389906,
      3.9872025288851036,
      0.17901102978375127,
      3.885476929518457,
      -0.8079039091377689,
      3.3940516818407187,
      -1.550651407188842,
      2.5792964886320684,
      -1.9489192990517052,
      1.5512485534497125,
      -1.9489192990517057,
      0.44875144655029087,
      -1.5506514071888438,
      -0.5792964886320653,
      -0.8079039091377715,
      -1.394051681840717,
      0.17901102978374794,
      -1.8854769295184561,
      1.276805078389902,
      -1.987202528885104,
      2.337215067329611,
      -1.6854898740651891,
      3.217026751661974,
      -1.021086930939675,
      3.7974166882130653,
      -0.08372499856146409,
    ]);
    validate(data);
  });

  test('corner case 4', () {
    final Float32List data = Float32List.fromList(<double>[
      -537.7739674441619,
      -122.26130468750004,
      -495.533967444162,
      -183.39195703125006,
      -453.29396744416204,
      -244.5226093750001,
      -411.0539674441621,
      -305.6532617187501,
      -164,
      -122,
    ]);
    validate(data);
  });

  test('returns empty triangulation for a small number of points', () {
    Delaunay d = Delaunay.from(<Point<double>>[])..update();
    expect(d.triangles, isEmpty);
    expect(d.hull, isEmpty);
    d = Delaunay.from(<Point<double>>[const Point<double>(0, 0)])..update();
    expect(d.triangles, isEmpty);
    expect(d.hull, isEmpty);
    d = Delaunay.from(<Point<double>>[
      const Point<double>(0, 0),
      const Point<double>(1, 1),
    ])
      ..update();
    expect(d.triangles, isEmpty);
    expect(d.hull, isEmpty);
  });

  test('returns empty triangulation for all colinear points', () {
    final Delaunay d = Delaunay(Float32List.fromList(<double>[
      0,
      0,
      1,
      0,
      3,
      0,
      2,
      0,
    ]))
      ..update();
    expect(d.triangles, isEmpty);
    expect(d.hull, equals(<int>[0, 1, 3, 2]));
  });
}

final double _epsilon = pow(2.0, -51) as double;

void validate(
  Float32List points, [
  Delaunay? d,
]) {
  d ??= Delaunay(points);
  d.update();

  // validate halfedges
  for (int i = 0; i < d.halfEdges.length; i++) {
    final int i2 = d.halfEdges[i];
    if (i2 != -1 && d.halfEdges[i2] != i) {
      fail('invalid halfedge connection');
    }
  }

  // validate triangulation
  final List<double> hullAreas = <double>[];
  final int len = d.hull.length;
  int prev = len - 1;
  for (int i = 0; i < len; i++) {
    final double x0 = d.coords[2 * d.hull[prev]];
    final double y0 = d.coords[2 * d.hull[prev] + 1];
    final double x = d.coords[2 * d.hull[i]];
    final double y = d.coords[2 * d.hull[i] + 1];

    hullAreas.add((x - x0) * (y + y0));
    final bool c = convex(
      d.coords[2 * d.hull[prev]],
      d.coords[2 * d.hull[prev] + 1],
      d.coords[2 * d.hull[(prev + 1) % d.hull.length]],
      d.coords[2 * d.hull[(prev + 1) % d.hull.length] + 1],
      d.coords[2 * d.hull[(prev + 3) % d.hull.length]],
      d.coords[2 * d.hull[(prev + 3) % d.hull.length] + 1],
    );
    if (!c) {
      fail('hull is not convex at $prev');
    }
    prev = i;
  }
  final double hullArea = sum(hullAreas);

  final List<double> triangleAreas = <double>[];
  for (int i = 0; i < d.triangles.length; i += 3) {
    final double ax = d.coords[2 * d.triangles[i]];
    final double ay = d.coords[2 * d.triangles[i] + 1];
    final double bx = d.coords[2 * d.triangles[i + 1]];
    final double by = d.coords[2 * d.triangles[i + 1] + 1];
    final double cx = d.coords[2 * d.triangles[i + 2]];
    final double cy = d.coords[2 * d.triangles[i + 2] + 1];
    triangleAreas.add(((by - ay) * (cx - bx) - (bx - ax) * (cy - by)).abs());
  }
  final double trianglesArea = sum(triangleAreas);

  final double err = ((hullArea - trianglesArea) / hullArea).abs();
  if (err > _epsilon) {
    fail('triangulation is broken: $err error');
  }
}

double orient(
  double px,
  double py,
  double rx,
  double ry,
  double qx,
  double qy,
) {
  final double l = (ry - py) * (qx - px);
  final double r = (rx - px) * (qy - py);
  return (l - r).abs() >= 3.3306690738754716e-16 * (l + r).abs() ? l - r : 0;
}

bool convex(
  double rx,
  double ry,
  double qx,
  double qy,
  double px,
  double py,
) {
  double orientation = orient(px, py, rx, ry, qx, qy);
  if (orientation != 0.0) {
    return orientation >= 0.0;
  }
  orientation = orient(rx, ry, qx, qy, px, py);
  if (orientation != 0.0) {
    return orientation >= 0.0;
  }
  orientation = orient(qx, qy, px, py, rx, ry);
  if (orientation != 0.0) {
    return orientation >= 0.0;
  }
  return true;
}

// Kahan and Babuska summation, Neumaier variant; accumulates less FP error
double sum(List<double> x) {
  double sum = x[0];
  double err = 0.0;
  for (int i = 1; i < x.length; i++) {
    final double k = x[i];
    final double m = sum + k;
    err += sum.abs() >= k.abs() ? sum - m + k : k - m + sum;
    sum = m;
  }
  return sum + err;
}


benchmark/delaunay_benchmark.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// A benchmark that depends only on this package and the core libraries to
// experiment with nnbd.

import 'dart:math';

import 'package:delaunay/delaunay.dart';

const int kSeed = 42;
const int kPoints = 1000000;
const int kWidth = 4096;
const int kHeight = 1024;

int main() {
  final Random r = Random(kSeed);
  final double maxX = kWidth.toDouble();
  final double maxY = kHeight.toDouble();

  final List<Point<double>> points = <Point<double>>[];
  for (int i = 0; i < kPoints; i++) {
    points.add(Point<double>(r.nextDouble() * maxX, r.nextDouble() * maxY));
  }

  final Delaunay triangulator = Delaunay.from(points);
  final Stopwatch sw = Stopwatch()..start();

  triangulator.initialize();
  print('Triangulator initialized in ${sw.elapsedMilliseconds}ms.');

  sw.reset();
  sw.start();

  triangulator.processAllPoints();
  print('Triangulated with ${triangulator.triangles.length ~/ 3} triangles '
      'in ${sw.elapsedMilliseconds}ms');
  return 0;
}


lib/src/delaunay_base.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math';
import 'dart:typed_data';

final double _epsilon = pow(2.0, -52) as double;
final Uint32List _edgeStack = Uint32List(512);

/// A class for quickly calculating the Delaunay triangulation in 2D of a set
/// of points in the cartesian plane.
///
/// Example usage:
/// ```
/// Float32List points = Float32List.fromList(<double>[
///   143.0, 178.5,
///   50.2, -100.7,
///   ...
/// ]);
/// Delaunay delaunay = Delaunay(points);
/// delaunay.update();
/// for (int i = 0; i < delaunay.triangles.length; i += 3) {
///   int a = delaunay.triangles[i];
///   int b = delaunay.triangles[i + 1];
///   int c = delaunay.triangles[i + 3];
///
///   double ax = delaunay.coords[2*a];
///   double ay = delaunay.coords[2*a + 1];
///   ...
/// }
/// ...
/// points[0] = 140.0;
/// delaunay.update();
/// ...
/// ```
///
/// This implementation is adapted from the Delaunator JavaScript library.
/// See https://github.com/mapbox/delaunator
class Delaunay {
  /// Allocates memory for a Delaunay triangulation, keeping a reference to
  /// [coords].
  Delaunay(Float32List coords)
      : _inputCoords = coords,

        // Provisionally set _coords to the input list. Decide later whether to
        // sort and replace.
        _coords = coords,

        // Arrays that will store the triangulation graph.
        _triangles = Uint32List(max(coords.length - 5, 0) * 3),
        _halfEdges = Int32List(max(coords.length - 5, 0) * 3),
        _trianglesLen = 0,
        _hullSize = 0,
        _hull = Uint32List(0),

        // Temporary arrays for tracking the edges of the advancing convex hull.
        _hullStart = -1,
        _hullPrev = Uint32List(coords.length >> 1),
        _hullNext = Uint32List(coords.length >> 1),
        _hullTri = Uint32List(coords.length >> 1),
        _hashSize = 1 << (sqrt(coords.length << 1).ceil().bitLength + 1),
        _hashMask = (1 << (sqrt(coords.length << 1).ceil().bitLength + 1)) - 1,
        _hullHash =
            Int32List(1 << (sqrt(coords.length << 1).ceil().bitLength + 1))
              ..fillRange(
                  0, 1 << (sqrt(coords.length << 1).ceil().bitLength + 1), -1),
        _i0 = -1,
        _i1 = -1,
        _i2 = -1,
        _cx = 0.0,
        _cy = 0.0,
        // Arrays to help sort the points.
        _ids = Uint32List(coords.length >> 1),
        _dists = Float32List(coords.length >> 1),
        _colinear = false;

  /// Allocates memory for a Delaunay triangulation.
  ///
  /// Copies [points] into a [Float32List], available from the getter [coords].
  factory Delaunay.from(List<Point<double>> points) {
    final int n = points.length;
    final Float32List coords = Float32List(n << 1);
    for (int i = 0; i < n; i++) {
      final Point<double> p = points[i];
      coords[i * 2] = p.x;
      coords[(i * 2) + 1] = p.y;
    }
    return Delaunay(coords);
  }

  /// The list of triangles in the triagulation.
  ///
  /// Each three elements are the three points of a triangle given by
  /// indices into [coords]:
  ///
  /// ```
  /// int a = triangles[i];
  /// int b = triangles[i + 1];
  /// int c = triangles[i + 2];
  ///
  /// double ax = coords[2 * a];
  /// double ay = coords[2 * a + 1];
  /// ```
  Uint32List get triangles => Uint32List.sublistView(
        _triangles,
        0,
        _trianglesLen,
      );

  /// The list of edges in the triangulation.
  Int32List get halfEdges => Int32List.sublistView(
        _halfEdges,
        0,
        _trianglesLen,
      );

  /// The coordinates being triangulated.
  ///
  /// Even indices are the x coordinates of the points, and the odd indices
  /// are the y coordinates of the points. E.g.:
  ///
  /// ```
  /// double px = coords[2*i];
  /// double py = coords[2*i + 1];
  /// ```
  Float32List get coords => _coords;

  /// The indices in [coords] of the points on the convex hull of the input
  /// data, counter-clockwise.
  Uint32List get hull => _hull;

  /// `coords[i]` in the form of a `Point<double>`.
  Point<double> getPoint(int i) => Point<double>(
        _coords[2 * i],
        _coords[2 * i + 1],
      );

  /// Whether after a call to [initialize], it has been detected that this is
  /// the degenerate case of all points being colinear.
  bool get colinear => _colinear;

  // The coordinates given to the constructor.
  final Float32List _inputCoords;

  // If needed for performance, the list of coordinates sorted in order of
  // increasing distance from the seed triangle's circumcenter.
  Float32List? _sortedCoords;

  // A reference either to _inputCoords or _coords whichever is indicated
  // by the number of points.
  Float32List _coords;

  // The triangles.
  final Uint32List _triangles;

  // The edges.
  final Int32List _halfEdges;

  final int _hashSize;
  final int _hashMask;
  final Int32List _hullHash;
  int _hullStart;
  int _hullSize;
  Uint32List _hull;
  final Uint32List _hullPrev;
  final Uint32List _hullNext;
  final Uint32List _hullTri;

  // Temporary arrays for sorting points.
  final Uint32List _ids;
  final Float32List _dists;

  int _i0, _i1, _i2;
  double _cx;
  double _cy;
  int _trianglesLen;

  bool _colinear;

  /// Calculates or recalculates the Delaunay triangulation.
  ///
  /// Calls [initialize] and [processAllPoints]. This can be called to
  /// perform the initial calculation, or to recalculate the triangulation
  /// if the contents of [coords] is modified.
  void update() {
    initialize();
    processAllPoints();
  }

  /// Initializes the Delaunay triangulation by finding the seed triangle and
  /// sorting the points in increasing distance from the seed triangle's
  /// circumcenter.
  void initialize() {
    final int n = _inputCoords.length >> 1;
    if (n < 3) {
      return;
    }

    // Populate an array of point indices; calculate input data bbox.
    double minX = double.maxFinite;
    double minY = double.maxFinite;
    double maxX = -double.maxFinite;
    double maxY = -double.maxFinite;
    for (int i = 0; i < n; i++) {
      final double x = _inputCoords[2 * i];
      final double y = _inputCoords[2 * i + 1];
      if (x < minX) {
        minX = x;
      }
      if (y < minY) {
        minY = y;
      }
      if (x > maxX) {
        maxX = x;
      }
      if (y > maxY) {
        maxY = y;
      }
      _ids[i] = i;
    }
    final double cx = (maxX + minX) / 2.0;
    final double cy = (maxY + minY) / 2.0;

    double minDist = double.maxFinite;
    int i0 = 0;
    int i1 = 0;
    int i2 = 0;

    // Pick a seed point close to the center.
    for (int i = 0; i < n; i++) {
      final double d = _dist(
        cx,
        cy,
        _inputCoords[2 * i],
        _inputCoords[2 * i + 1],
      );
      if (d < minDist) {
        i0 = i;
        minDist = d;
      }
    }
    final double i0x = _inputCoords[2 * i0];
    final double i0y = _inputCoords[2 * i0 + 1];

    // Find the point closest to the seed.
    minDist = double.maxFinite;
    for (int i = 0; i < n; i++) {
      if (i == i0) {
        continue;
      }
      final double d = _dist(
        i0x,
        i0y,
        _inputCoords[2 * i],
        _inputCoords[2 * i + 1],
      );
      if (d < minDist && d > 0.0) {
        i1 = i;
        minDist = d;
      }
    }
    double i1x = _inputCoords[2 * i1];
    double i1y = _inputCoords[2 * i1 + 1];

    // Find the third point which forms the smallest circumcircle with the first
    // two.
    double minRadius = double.maxFinite;
    for (int i = 0; i < n; i++) {
      if (i == i0 || i == i1) {
        continue;
      }
      final double r = _circumradius(
        i0x,
        i0y,
        i1x,
        i1y,
        _inputCoords[2 * i],
        _inputCoords[2 * i + 1],
      );
      if (r < minRadius) {
        i2 = i;
        minRadius = r;
      }
    }
    double i2x = _inputCoords[2 * i2];
    double i2y = _inputCoords[2 * i2 + 1];

    if (minRadius == double.maxFinite) {
      // Order collinear points by dx (or dy if all x are identical)
      // and return the list as a hull
      for (int i = 0; i < n; i++) {
        final double dx = _inputCoords[2 * i] - _inputCoords[0];
        _dists[i] = dx != 0.0 ? dx : _inputCoords[2 * i + 1] - _inputCoords[1];
      }
      _quicksort(_ids, _dists, 0, n - 1);
      final Uint32List hull = Uint32List(n);
      int j = 0;
      double d0 = -double.maxFinite;
      for (int i = 0; i < n; i++) {
        final int id = _ids[i];
        if (_dists[id] > d0) {
          hull[j++] = id;
          d0 = _dists[id];
        }
      }
      _hull = hull.sublist(0, j);
      _colinear = true;
      return;
    }

    // Swap the order of the seed points for counter-clockwise orientation.
    if (_orient(i0x, i0y, i1x, i1y, i2x, i2y)) {
      final int i = i1;
      final double x = i1x;
      final double y = i1y;
      i1 = i2;
      i1x = i2x;
      i1y = i2y;
      i2 = i;
      i2x = x;
      i2y = y;
    }

    // Sort the points by distance from the seed triangle circumcenter
    final Point<double> center = _circumcenter(i0x, i0y, i1x, i1y, i2x, i2y);
    _cx = center.x;
    _cy = center.y;
    for (int i = 0; i < n; i++) {
      _dists[i] = _dist(_inputCoords[2 * i], _inputCoords[2 * i + 1], _cx, _cy);
    }
    _quicksort(_ids, _dists, 0, n - 1);

    if (n <= 100000) {
      _i0 = i0;
      _i1 = i1;
      _i2 = i2;
      _coords = _inputCoords;
    } else {
      // When there are more than ~100000 points, improving the locality of
      // access to the coordinate list improves performance.
      _sortedCoords ??= Float32List(_inputCoords.length);
      final Float32List localSorted = _sortedCoords!;
      for (int i = 0; i < n; i++) {
        if (_ids[i] == i0) {
          _i0 = i;
        }
        if (_ids[i] == i1) {
          _i1 = i;
        }
        if (_ids[i] == i2) {
          _i2 = i;
        }
        localSorted[2 * i] = _inputCoords[2 * _ids[i]];
        localSorted[2 * i + 1] = _inputCoords[2 * _ids[i] + 1];
        _ids[i] = i;
      }
      _coords = localSorted;
    }

    // Set up the seed triangle as the starting hull.
    _hullStart = _i0;
    _hullSize = 3;
    _hullNext[_i0] = _hullPrev[_i2] = _i1;
    _hullNext[_i1] = _hullPrev[_i0] = _i2;
    _hullNext[_i2] = _hullPrev[_i1] = _i0;

    _hullTri[_i0] = 0;
    _hullTri[_i1] = 1;
    _hullTri[_i2] = 2;

    _hullHash.fillRange(0, _hullHash.length, -1);
    _hullHash[_hashKey(i0x, i0y)] = _i0;
    _hullHash[_hashKey(i1x, i1y)] = _i1;
    _hullHash[_hashKey(i2x, i2y)] = _i2;

    _trianglesLen = 0;
    _addTriangle(_i0, _i1, _i2, -1, -1, -1);
  }

  /// Iterates over all the points in order of increasing distance from the seed
  /// triangle's circumcenter, adding them to the triangulation.
  ///
  /// After this call, the getters [triangles], [halfEdges], and [hull] will
  /// have meaningful contents.
  void processAllPoints() {
    final int n = _coords.length >> 1;
    if (n < 3 || _colinear) {
      return;
    }
    double xp = 0.0, yp = 0.0;
    for (int i = 0; i < n; i++) {
      final int idx = _ids[i];
      final double x = _coords[2 * idx];
      final double y = _coords[2 * idx + 1];

      if (i > 0 && (x - xp).abs() <= _epsilon && (y - yp).abs() <= _epsilon) {
        // Too close to the previous point. Skip it.
        continue;
      }
      xp = x;
      yp = y;

      // Skip seed triangle points.
      if (idx == _i0 || idx == _i1 || idx == _i2) {
        continue;
      }

      _hullSize = _processNextPoint(_hullSize, idx, x, y);
    }

    _hull = Uint32List(_hullSize);
    int e = _hullStart;
    for (int i = 0; i < _hullSize; i++) {
      _hull[i] = e;
      e = _hullNext[e];
    }
  }

  int _processNextPoint(int hullSize, int i, double x, double y) {
    // Find a visible edge on the convex hull using edge hash.
    int start = 0;
    final int key = _hashKey(x, y);
    for (int j = 0; j < _hashSize; j++) {
      start = _hullHash[(key + j) & _hashMask];
      if (start != -1 && start != _hullNext[start]) {
        break;
      }
    }

    start = _hullPrev[start];
    int e = start;
    int q;
    while (true) {
      q = _hullNext[e];
      final bool orient = _orient(
        x,
        y,
        _coords[2 * e],
        _coords[2 * e + 1],
        _coords[2 * q],
        _coords[2 * q + 1],
      );
      if (orient) {
        break;
      }
      e = q;
      if (e == start) {
        e = -1;
        break;
      }
    }
    if (e == -1) {
      // Likely a near-duplicate point; skip it.
      return hullSize;
    }

    // Add the first triangle from the point.
    int t = _addTriangle(e, i, _hullNext[e], -1, -1, _hullTri[e]);

    // Recursively flip triangles from the point until they satisfy the
    // Delaunay condition.
    _hullTri[i] = _legalize(t + 2);
    _hullTri[e] = t; // Keep track of boundary triangles on the hull.
    hullSize++;

    // Walk forward through the hull, adding more triangles and flipping
    // recursively.
    int n = _hullNext[e];
    while (true) {
      q = _hullNext[n];
      final bool orient = _orient(
        x,
        y,
        _coords[2 * n],
        _coords[2 * n + 1],
        _coords[2 * q],
        _coords[2 * q + 1],
      );
      if (!orient) {
        break;
      }
      t = _addTriangle(n, i, q, _hullTri[i], -1, _hullTri[n]);
      _hullTri[i] = _legalize(t + 2);
      _hullNext[n] = n; // Mark as removed.
      hullSize--;
      n = q;
    }

    // Walk backward from the other side, adding more triangles and flipping.
    if (e == start) {
      while (true) {
        q = _hullPrev[e];
        final bool orient = _orient(
          x,
          y,
          _coords[2 * q],
          _coords[2 * q + 1],
          _coords[2 * e],
          _coords[2 * e + 1],
        );
        if (!orient) {
          break;
        }
        t = _addTriangle(q, i, e, -1, _hullTri[e], _hullTri[q]);
        _legalize(t + 2);
        _hullTri[q] = t;
        _hullNext[e] = e; // Mark as removed.
        hullSize--;
        e = q;
      }
    }

    // Update the hull indices.
    _hullStart = _hullPrev[i] = e;
    _hullNext[e] = _hullPrev[n] = i;
    _hullNext[i] = n;

    // Save the two new edges in the hash table.
    _hullHash[_hashKey(x, y)] = i;
    _hullHash[_hashKey(_coords[2 * e], _coords[2 * e + 1])] = e;
    return hullSize;
  }

  int _hashKey(double x, double y) {
    return (_pseudoAngle(x - _cx, y - _cy) * _hashSize).floor() & _hashMask;
  }

  // Monotonically increases with real angle, but doesn't need expensive
  // trigonometry. Search for 'Diamond Angle'.
  double _pseudoAngle(double dx, double dy) {
    final double p = dx / (dx.abs() + dy.abs());
    return (dy > 0.0 ? 3.0 - p : 1.0 + p) / 4.0; // [0..1]
  }

  // We know that the indices will be no bigger than 32-bit ints, so we
  // perform 'mod 3' using the 32-bit multiplicative inverse instead of
  // division, which is much slower. See, among others:
  // https://stackoverflow.com/questions/44212301/efficient-mod-3-in-x86-assembly
  int _mod3(int a) {
    final int r = a * 0xaaaaaaab;
    final int q = r >> 33;
    final int q3 = (q << 1) + q;
    return a - q3;
  }

  int _legalize(int a) {
    int i = 0;
    int ar = 0;

    // Recursion eliminated with a fixed-size stack.
    while (true) {
      final int b = _halfEdges[a];

      // If the pair of triangles doesn't satisfy the Delaunay condition
      // (p1 is inside the circumcircle of [p0, pl, pr]), flip them,
      // then do the same check/flip recursively for the new pair of triangles.
      //
      //           pl                    pl
      //          /||\                  /  \
      //       al/ || \bl            al/    \a
      //        /  ||  \              /      \
      //       /  a||b  \    flip    /___ar___\
      //     p0\   ||   /p1   =>   p0\---bl---/p1
      //        \  ||  /              \      /
      //       ar\ || /br             b\    /br
      //          \||/                  \  /
      //           pr                    pr
      //
      final int a0 = a - _mod3(a);
      ar = a0 + _mod3(a + 2);

      if (b == -1) {
        // Convex hull edge.
        if (i == 0) {
          break;
        }
        a = _edgeStack[--i];
        continue;
      }

      final int b0 = b - _mod3(b);
      final int al = a0 + _mod3(a + 1);
      final int bl = b0 + _mod3(b + 2);

      final int p0 = _triangles[ar];
      final int pr = _triangles[a];
      final int pl = _triangles[al];
      final int p1 = _triangles[bl];

      bool illegal;
      {
        // Check whether p1 is in the circumcircle of p0, pr, and pl.
        final double ax = _coords[2 * p0];
        final double ay = _coords[2 * p0 + 1];
        final double bx = _coords[2 * pr];
        final double by = _coords[2 * pr + 1];
        final double cx = _coords[2 * pl];
        final double cy = _coords[2 * pl + 1];
        final double px = _coords[2 * p1];
        final double py = _coords[2 * p1 + 1];
        final double dx = ax - px;
        final double dy = ay - py;
        final double ex = bx - px;
        final double ey = by - py;
        final double fx = cx - px;
        final double fy = cy - py;

        final double ap = dx * dx + dy * dy;
        final double bp = ex * ex + ey * ey;
        final double cp = fx * fx + fy * fy;

        illegal = dx * (ey * cp - bp * fy) -
                dy * (ex * cp - bp * fx) +
                ap * (ex * fy - ey * fx) <
            0.0;
      }

      if (illegal) {
        _triangles[a] = p1;
        _triangles[b] = p0;

        final int hbl = _halfEdges[bl];

        // Edge swapped on the other side of the hull (rare); fix the halfedge
        // reference.
        if (hbl == -1) {
          int e = _hullStart;
          do {
            if (_hullTri[e] == bl) {
              _hullTri[e] = a;
              break;
            }
            e = _hullPrev[e];
          } while (e != _hullStart);
        }
        _link(a, hbl);
        _link(b, _halfEdges[ar]);
        _link(ar, bl);

        final int br = b0 + _mod3(b + 1);

        // Don't worry about hitting the cap: it can only happen on extremely
        // Degenerate input.
        if (i < _edgeStack.length) {
          _edgeStack[i++] = br;
        }
      } else {
        if (i == 0) {
          break;
        }
        a = _edgeStack[--i];
      }
    }

    return ar;
  }

  double _circumradius(
    double ax,
    double ay,
    double bx,
    double by,
    double cx,
    double cy,
  ) {
    final double dx = bx - ax;
    final double dy = by - ay;
    final double ex = cx - ax;
    final double ey = cy - ay;

    final double bl = dx * dx + dy * dy;
    final double cl = ex * ex + ey * ey;
    final double d = 0.5 / (dx * ey - dy * ex);

    final double x = (ey * bl - dy * cl) * d;
    final double y = (dx * cl - ex * bl) * d;

    return x * x + y * y;
  }

  Point<double> _circumcenter(
    double ax,
    double ay,
    double bx,
    double by,
    double cx,
    double cy,
  ) {
    final double dx = bx - ax;
    final double dy = by - ay;
    final double ex = cx - ax;
    final double ey = cy - ay;

    final double bl = dx * dx + dy * dy;
    final double cl = ex * ex + ey * ey;
    final double d = 0.5 / (dx * ey - dy * ex);

    final double x = ax + (ey * bl - dy * cl) * d;
    final double y = ay + (dx * cl - ex * bl) * d;

    return Point<double>(x, y);
  }

  void _link(int a, int b) {
    _halfEdges[a] = b;
    if (b != -1) {
      _halfEdges[b] = a;
    }
  }

  // Add a new triangle given vertex indices and adjacent half-edge ids.
  int _addTriangle(int i0, int i1, int i2, int a, int b, int c) {
    final int t = _trianglesLen;

    _triangles[t] = i0;
    _triangles[t + 1] = i1;
    _triangles[t + 2] = i2;

    _link(t, a);
    _link(t + 1, b);
    _link(t + 2, c);

    _trianglesLen += 3;

    return t;
  }

  double _dist(double ax, double ay, double bx, double by) {
    final double dx = ax - bx;
    final double dy = ay - by;
    return dx * dx + dy * dy;
  }

  // Return 2d orientation sign if we're confident in it through J. Shewchuk's
  // error bound check.
  double _orientIfSure(
    double px,
    double py,
    double rx,
    double ry,
    double qx,
    double qy,
  ) {
    final double l = (ry - py) * (qx - px);
    final double r = (rx - px) * (qy - py);
    return (l - r).abs() >= 3.3306690738754716e-16 * (l + r).abs()
        ? l - r
        : 0.0;
  }

  // A more robust orientation test that's stable in a given triangle (to fix
  // robustness issues).
  bool _orient(
    double rx,
    double ry,
    double qx,
    double qy,
    double px,
    double py,
  ) {
    double orientation = _orientIfSure(px, py, rx, ry, qx, qy);
    if (orientation != 0.0) {
      return orientation < 0.0;
    }
    orientation = _orientIfSure(rx, ry, qx, qy, px, py);
    if (orientation != 0.0) {
      return orientation < 0.0;
    }
    orientation = _orientIfSure(qx, qy, px, py, rx, ry);
    if (orientation != 0.0) {
      return orientation < 0.0;
    }
    return true;
  }

  void _quicksort(Uint32List ids, Float32List dists, int left, int right) {
    if (right - left <= 20) {
      for (int i = left + 1; i <= right; i++) {
        final int temp = ids[i];
        final double tempDist = dists[temp];
        int j = i - 1;
        while (j >= left && dists[ids[j]] > tempDist) {
          ids[j + 1] = ids[j--];
        }
        ids[j + 1] = temp;
      }
    } else {
      final int median = (left + right) >> 1;
      int i = left + 1;
      int j = right;
      swap(ids, median, i);
      if (dists[ids[left]] > dists[ids[right]]) {
        swap(ids, left, right);
      }
      if (dists[ids[i]] > dists[ids[right]]) {
        swap(ids, i, right);
      }
      if (dists[ids[left]] > dists[ids[i]]) {
        swap(ids, left, i);
      }

      final int temp = ids[i];
      final double tempDist = dists[temp];
      while (true) {
        do {
          i++;
        } while (dists[ids[i]] < tempDist);
        do {
          j--;
        } while (dists[ids[j]] > tempDist);
        if (j < i) {
          break;
        }
        swap(ids, i, j);
      }
      ids[left + 1] = ids[j];
      ids[j] = temp;

      if (right - i + 1 >= j - left) {
        _quicksort(ids, dists, i, right);
        _quicksort(ids, dists, left, j - 1);
      } else {
        _quicksort(ids, dists, left, j - 1);
        _quicksort(ids, dists, i, right);
      }
    }
  }

  void swap(Uint32List arr, int i, int j) {
    final int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }
}
